index.bs

```diff
@@ -2756,6 +2756,70 @@ because use of {{ScriptProcessorNode}} from the main thread
 frequently resulted in a poor user experience. [[WebAudio]]
 </p>
 
+## Only purely computational features should be exposed everywhere ## {#expose-everywhere}
+<!--
+    https://github.com/w3ctag/design-principles/issues/509
+    https://github.com/whatwg/webidl/issues/468
+    https://github.com/w3ctag/design-principles/issues/35
+-->
+
+When exposing a feature,
+please consider whether it makes sense
+to expose the feature to all possible environments
+(via the {{Exposed|[Exposed=*]}} annotation
+or including it on all global scope interfaces).
+
+Only purely computational features should be exposed everywhere.
+That is,
+they do not perform I/O
+and do not affect the state of the user agent or the user’s device.
+
+<p class="example">
+The {{TextEncoder}} interface converts a string to UTF-8 encoded bytes.
+This is a purely computational interface,
+generally useful as a JavaScript language facility,
+so it should be exposed everywhere.
+</p>
+
+<p class="example">
+<a attribute spec=html>localStorage</a> affects the state of the user agent, so it should not be exposed everywhere.
+</p>
+
+<p class="example">
+Technically, {{console}} could affect the state of the user agent
+(by causing log messages to appear in the developer tools)
+or the user’s device
+(by writing to a log file.)
+But these things are not observable from the running code,
+and the practicality of having {{console}} everywhere outweighs the disadvantages.
+</p>
+
+Additionally, anything annotated with {{SecureContext}},
+or relying on an event loop,
+should not be exposed everywhere.
+Not all global scopes are secure contexts
+or have an event loop.
```

Comment:
I'm not sure about excluding [`[SecureContext]`](https://webidl.spec.whatwg.org/#SecureContext) things from `[Exposed=*]`. Won't `[SecureContext]` itself exclude those items from global scopes where they'd be inappropriate?

```suggestion
Additionally, anything relying on an event loop
should not be exposed everywhere.
Not all global scopes have an event loop.
```

Reply:
Thanks for the review!

My motivation for this was to get exactly the same set of exposed APIs in ShadowRealm regardless of which global scope created the ShadowRealm. I could go either way on this; we could remove this sentence from the guidance and specify in https://github.com/whatwg/html/pull/9893 that ShadowRealm scopes are never secure contexts. That would have the same result, I think.

Reply:
Since browsers are going to differ in which APIs they implement anyway, I don't think there's that much reason to worry about getting exactly the same set of exposed APIs in ShadowRealm. By contrast, I do think it's important that pure-computation APIs which happen to be restricted to secure contexts, like [`crypto.randomUUID()`](https://github.com/WICG/uuid/issues/23), should be available in ShadowRealms when they are themselves running in secure contexts.

The stated intention of browsers in restricting features to secure-context-only is to encourage the use of secure contexts for all new applications, with this being assumed (IMO incorrectly) not to be a very large cost to the usability of the feature because applications should be in secure contexts anyway so developers following best practices (i.e. secure contexts) shouldn't even notice. That will not be true if we have this restriction here.

Reply:
I'm not sure I agree, but I will at least remove `[SecureContext]` from here so that discussion becomes irrelevant to this PR and we can continue talking about whether a ShadowRealm is a secure context in the HTML PR.